File: 4.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <regex>
#include <algorithm>
#include <unordered_map>
#include <set>

using namespace std;
using my_key_t = string;
using my_val_t = string;
using passport_field_value_t = unordered_map<my_key_t,my_val_t>;

const string inputname = "4.in";

bool is_numeric(string s)
{
    return !s.empty() && s.find_first_not_of("0123456789") == std::string::npos;
}

bool validate_byr(string x)
{
    return (x.size()==4) && is_numeric(x) && ((1920<=stoi(x))&&(stoi(x)<=2002));
}

bool validate_iyr(string x)
{
    return (x.size()==4) && is_numeric(x) && ((2010<=stoi(x))&&(stoi(x)<=2020));
}

bool validate_eyr(string x)
{
    return (x.size()==4) && is_numeric(x) && ((2020<=stoi(x))&&(stoi(x)<=2030));
}

bool validate_hgt(string x)
{
    bool valid = false;
    if ((x.size()==4) or (x.size()==5))
    {
        int val = stoi(x.substr(0, x.size()-2));
        if (x.find("in") != std::string::npos)
        {
            valid = (59 <= val) && (val <= 76);
        }
        else if (x.find("cm") != std::string::npos)
        {
            valid = (150 <= val) && (val <= 193);
        }
    }
    return valid;
}

bool validate_hcl(string x)
{
    return regex_search(x,regex{"#([a-f0-9]{6})"});
}

bool validate_ecl(string x)
{
    std::set<string> colours{"amb","blu","brn","gry","grn","hzl","oth"};
    return colours.find(x) != colours.end();
}

bool validate_pid(string x)
{
    return (x.size()==9) && (is_numeric(x));
}

bool validate_cid(string x)
{
    return true;
}


typedef bool (*validator_fcn_t)(string);

unordered_map<string, validator_fcn_t> val_fun = {
    {"byr", validate_byr},
    {"iyr", validate_iyr},
    {"eyr", validate_eyr},
    {"hgt", validate_hgt},
    {"hcl", validate_hcl},
    {"ecl", validate_ecl},
    {"pid", validate_pid},
    {"cid", validate_cid},
    };

int one(vector<passport_field_value_t> & ls)
{
    int valid = 0;
    for (auto & d: ls)
    {
        std::vector<my_key_t> keys;
        keys.reserve(d.size());
        std::vector<my_val_t> vals;
        vals.reserve(d.size());

        for(auto & kv : d)
        {
            keys.push_back(kv.first);
            vals.push_back(kv.second);
        }
        
        if ((keys.size() == 8) ||
            ((keys.size() == 7) && ( std::find(keys.begin(), keys.end(), "cid") == keys.end() )))
        {
            valid += 1;
        }
    }
    return (valid);
}

int two(vector<passport_field_value_t> & ls)
{
    int valid = 0;
    for (auto & d: ls)
    {
        std::vector<my_key_t> keys;
        keys.reserve(d.size());
        std::vector<my_val_t> vals;
        vals.reserve(d.size());

        for(auto & kv : d)
        {
            keys.push_back(kv.first);
            vals.push_back(kv.second);
        }
        
        if ((keys.size() == 8) ||
            ((keys.size() == 7) && ( std::find(keys.begin(), keys.end(), "cid") == keys.end() )))
        {
            bool valid_flag = true;
            for(auto & kv : d)
            {
                valid_flag &= val_fun[kv.first](kv.second);
            }
            if (valid_flag)
            {
                ++valid;
            }
        }
    }
    return (valid);
}


int main(int argc, char **kwargs)
{
    ifstream infile(inputname);
    std::string chars = "\n\r";
    string s;
    passport_field_value_t d;
    vector<passport_field_value_t> ls;
    regex e("(\\S+):(\\S+)");
    while( getline( infile, s ) ) // for each line read from the file
    {
        // default constructor = end-of-sequence:
        std::regex_token_iterator<std::string::iterator> rend;

        int submatches[] = { 1, 2 };
        std::regex_token_iterator<std::string::iterator> c ( s.begin(), s.end(), e, submatches );
        // Empty line, start a new dictionary
        if (c==rend)
        {
            ls.push_back(d);
            d = {};
            continue;
        }
        while (c!=rend)
        {
            auto k = *c++;
            auto v = *c++;
            d[k] = v;
        }
    }
    // Last dictionary not yet added
    ls.push_back(d);
    
    std::cout << "one = " << one(ls) << '\n';
    std::cout << "two = " << two(ls) << '\n';

    return 0;
}

File: 4.py

import re
ls = []
d = {}

for s in open('4.in','r').readlines():
    pattern = "(\S+):(\S+)"
    match = re.findall(pattern, s)
    for M in match:
        k = M[0]
        v = M[1]
        d[k] = v
    if (s=='\n'):
        ls.append(d)
        d = {}
ls.append(d)

def validate_byr(x):
    return (len(x)==4) and (x.isnumeric()) and (1920<=int(x)<=2002)
    
def validate_iyr(x):
    return (len(x)==4) and (x.isnumeric()) and (2010<=int(x)<=2020)
    
def validate_eyr(x):
    return (len(x)==4) and (x.isnumeric()) and (2020<=int(x)<=2030)
    
def validate_hgt(x):
    valid = False
    if ((len(x)==4) or (len(x)==5)):
        if 'in' in x:
            valid = (59 <= int(x.rstrip('in')) <= 76)
        elif 'cm' in x:
            valid = (150 <= int(x.rstrip('cm')) <= 193)
        else:
            valid = False
    else:
        valid = False
    return valid
    
def validate_hcl(x):
    pattern = "#([a-f0-9]{6})"
    match = re.search(pattern, x)
    return match != None
    
def validate_ecl(x):
    return x in {'amb','blu','brn','gry','grn','hzl','oth'}
    
def validate_pid(x):
    return (len(x)==9) and x.isnumeric()

def validate_cid(x):
    return True

val_fun = { 'byr' : validate_byr,
            'iyr' : validate_iyr,
            'eyr' : validate_eyr,
            'hgt' : validate_hgt,
            'hcl' : validate_hcl,
            'ecl' : validate_ecl,
            'pid' : validate_pid,
            'cid' : validate_cid
           }

def one(ls):
    valid_passports = 0
    for d in ls:
        if ((len(d.keys()) == 8) or 
            ((len(d.keys()) == 7) and ('cid' not in d))):
            valid_passports += 1
    return valid_passports

def two(ls):
    valid_passports = 0
    for d in ls:
        valid = False
        if ((len(d.keys()) == 8) or 
            ((len(d.keys()) == 7) and ('cid' not in d))):
            valid = True
            for k,v in d.items():
                valid &= val_fun[k](v)
        if valid:
            valid_passports += 1
    return valid_passports
    
if __name__ == "__main__":
    print(f"{one(ls) = }") # 228
    print(f"{two(ls) = }") # 175
